iwyu-check-license-header.py

```diff
@@ -0,0 +1,266 @@
+        if not line or line.startswith('#!'):
+            continue
+
+        # Besides those, the first line should be a license header.
+        if line.startswith(two + '==='):
+            break
+
+        # If not, this fails the test entirely.
+        return -1
+
+    return i
+
+
+class File(object):
+    """ Base class for a source file with a license header
+
+    Do not use directly, instead use File.parse to instantiate a more derived
+    class.
+
+    Derived classes must have three class variables:
+
+    * one - One comment char ('#' for Python, '//' for C++)
+    * two - Two comment chars ('##' for Python, '//' for C++)
+    * pattern - a regex matching the first line in a license header
+    """
+    def __init__(self, filename):
+        with open(filename, 'r') as fd:
+            content = fd.read()
+
+        self.lines = list(content.splitlines())
+        self.hdrindex = find_license_header(self.lines, self.two)
+        self.filename = filename
+        self.errors = []
+
+    @classmethod
+    def parse(_, filename):
+        """ Return an object derived from File to analyze license headers """
+        _, ext = os.path.splitext(filename)
+        if ext == '.py':
+            klass = PythonFile
+        elif ext in ('.h', '.c', '.cc'):
+            klass = CxxFile
+        else:
+            return None
+
+        return klass(filename)
+
+    def has_license_header(self):
+        """ Return True if a license header has been found """
+        return self.hdrindex != -1
+
+    def add_license_header(self):
+        """ Add license header to a file that doesn't have one. """
+        assert not self.has_license_header()
+
+        # Find insertion point
+        for p, line in enumerate(self.lines):
+            # Skip past leading blank lines and hash-bangs.
+            if line and not line.startswith('#!'):
+                break
+
+        # Split the lines around the insertion point
+        if self.lines:
+            before, after = self.lines[:p], self.lines[p:]
+        else:
+            before, after = [], []
+
+        # Rebuild the contents with the license header in the middle
+        lines = before
+        if before and before[-1] != '':
+            lines += ['']
+        lines += make_license_header(self.filename, self.one, self.two)
+        if after and after[0] != '':
+            lines += ['']
+        lines += after
+
+        # Write back out
+        with open(self.filename, 'wb') as fd:
+            fd.write('\n'.join(lines))
+            fd.write('\n')
+
+    def check_license_header(self):
+        """ Check that the header lines follow convention.
+
+        Returns True if everything is OK, otherwise returns False and populates
+        self.errors with all found errors.
+        """
+        if not self.has_license_header():
+            self.file_error('No license header found')
+            return False
+
+        hdrlines = self.lines[self.hdrindex:self.hdrindex+8]
```

Comment:
Not likely to change but the magic number can be easily calculated.

Reply:
I find Python slicing with variables becomes very hard to read, so I'd prefer to keep the magic.